gnome-image-installer/pages/install/gis-scribe.c

```diff
@@ -1507,11 +1667,32 @@ gis_scribe_write_async (GisScribe          *self,
       goto out;
     }
 
-  /* Attempt to spawn GPG subprocess */
+  /* Attempt to spawn GPG subprocess or checksum thread */
   self->outstanding_tasks |= GIS_SCRIBE_TASK_VERIFY;
-  gpg_stdin = gis_scribe_begin_verify (self, cancellable, gis_scribe_subtask_cb,
-                                       g_object_ref (task));
-  if (gpg_stdin == NULL)
+  if (g_file_query_exists (self->signature, cancellable))
+    {
+      verify_pipe = gis_scribe_begin_verify_gpg (self, cancellable,
+                                                 gis_scribe_subtask_cb,
+                                                 g_object_ref (task));
+    }
+  else if (g_file_query_exists (self->checksum, cancellable))
+    {
+      verify_pipe = gis_scribe_begin_verify_checksum (self, cancellable,
+                                                      gis_scribe_subtask_cb,
+                                                      g_object_ref (task));
+    }
+  else
+    {
+      g_autofree gchar *signature_path = g_file_get_path (self->signature);
+      g_autofree gchar *checksum_path = g_file_get_path (self->checksum);
+      g_task_return_new_error (
+          task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("Neither the signature file ‘%s’ nor the checksum file ‘%s’ exist."),
+          signature_path, checksum_path);
+      goto out;
```

Comment:
Won't this skip the block just below that closes the other pipes?

Reply:
Oh, good point. I don't think I fully understand the error handling here.